### DISCUSSION TITLE
renovate-digest-bug-report

### DISCUSSION THREAD

**Discussion by Marvin-Highcreek** [renovatebot/renovate#31378] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

gitlab v16.11.6, renovate v38.21.3 

### Please tell us more about your question or problem

# renovate-digest-bug-report

When wusing the aws-machine-image manager I encountered a Bug. 

Using groupSlug to create dynamic branch names for our ami deployment instead of correctly using "renovate/all-ami-updates-dev" it falls back to this branch name: "renovate/ami_id-1.x."

From the documentation here https://docs.renovatebot.com/configuration-options/#branchtopic I can see how the branch name is being concatenated. 

![CleanShot 2024-09-13 at 13 45 40](https://github.com/user-attachments/assets/2b0ca0e2-4821-4bd5-a052-3a9568e2ff54)

"{{{depNameSanitized}}}-{{{newMajor}}}{{#if separateMinorPatch}}{{#if isPatch}}.{{{newMinor}}}{{/if}}{{/if}}.x{{#if isLockfileUpdate}}-lockfile{{/if}}"

When looking at our logs I can see this. 

```
{
  "bucket": "non-major",
  "newVersion": "ami-123",
  "newValue": "ami-123",
  "newDigest": "amazon-eks-node-1.29-vxxx",
  "releaseTimestamp": "XXX",
  "newMajor": 1,
  "newMinor": 0,
  "newPatch": 0,
  "updateType": "patch",
  "branchName": "renovate/ami_id-1.x"
}
```

So it is becoming clear to me why the branchname is named like it is but why the groupslug gets ignored is beyond me. 

To mitigate this, because we really need different branch names, I did this. 

```
{
  packageRules: [
    {
      matchPackageNames: ["ami-*"],
      groupName: `all AMI updates for ${process.env.AWS_ACCOUNT}`,
      groupSlug: `all-ami-updates-${process.env.AWS_ACCOUNT}`,
      branchNameStrict: "true",
      branchTopic: "{{{depNameSanitized}}}-{{{newDigest}}}-{{{newValue}}}",
    }
  ],
}
```
The result was still "branchName": "renovate/ami_id-1.x" .

When removing the {{{newDigest}}} part this solution works, so writing 

```
{
  packageRules: [
    {
      matchPackageNames: ["ami-*"],
      groupName: `all AMI updates for ${process.env.AWS_ACCOUNT}`,
      groupSlug: `all-ami-updates-${process.env.AWS_ACCOUNT}`,
      branchNameStrict: "true",
      branchTopic: "{{{depNameSanitized}}}-{{{newValue}}}",
    }
  ],
}
```
works and I get "branchName": "renovate/ami_id-ami-123".

Also placing branchTopic outside the package rules works

```
module.exports = {
  ...baseConfig,
  branchTopic: "{{{depNameSanitized}}}-{{{newDigest}}}-{{{newValue}}}",

packageRules: [
    {
      matchPackageNames: ["ami-*"],
      groupName: `all AMI updates for ${process.env.AWS_ACCOUNT}`,
      groupSlug: `all-ami-updates-${process.env.AWS_ACCOUNT}`,
      branchNameStrict: "true",
    }
  ],
}
```
and i even get the digest into the branch name like = "branchName": "renovate/ami_id-amazon-eks-node-1.29-vxxxx-xxxx-ami-123"

So it feels like there is an bug/error in how the digest gets read and renovate reverts to the default wich doesnt include the patch version.
# renovate-digest-bug-report

When wusing the aws-machine-image manager I encountered a Bug. 

Using groupSlug to create dynamic branch names for our ami deployment instead of correctly using "renovate/all-ami-updates-dev" it falls back to this branch name: "renovate/ami_id-1.x."

From the documentation here https://docs.renovatebot.com/configuration-options/#branchtopic I can see how the branch name is being concatenated. 

![CleanShot 2024-09-13 at 13 45 40](https://github.com/user-attachments/assets/2b0ca0e2-4821-4bd5-a052-3a9568e2ff54)

"{{{depNameSanitized}}}-{{{newMajor}}}{{#if separateMinorPatch}}{{#if isPatch}}.{{{newMinor}}}{{/if}}{{/if}}.x{{#if isLockfileUpdate}}-lockfile{{/if}}"

When looking at our logs I can see this. 

```
{
  "bucket": "non-major",
  "newVersion": "ami-123",
  "newValue": "ami-123",
  "newDigest": "amazon-eks-node-1.29-vxxx",
  "releaseTimestamp": "XXX",
  "newMajor": 1,
  "newMinor": 0,
  "newPatch": 0,
  "updateType": "patch",
  "branchName": "renovate/ami_id-1.x"
}
```

So it is becoming clear to me why the branchname is named like it is but why the groupslug gets ignored is beyond me. 

To mitigate this, because we really need different branch names, I did this. 

```
{
  packageRules: [
    {
      matchPackageNames: ["ami-*"],
      groupName: `all AMI updates for ${process.env.AWS_ACCOUNT}`,
      groupSlug: `all-ami-updates-${process.env.AWS_ACCOUNT}`,
      branchNameStrict: "true",
      branchTopic: "{{{depNameSanitized}}}-{{{newDigest}}}-{{{newValue}}}",
    }
  ],
}
```
The result was still "branchName": "renovate/ami_id-1.x" .

When removing the {{{newDigest}}} part this solution works, so writing 

```
{
  packageRules: [
    {
      matchPackageNames: ["ami-*"],
      groupName: `all AMI updates for ${process.env.AWS_ACCOUNT}`,
      groupSlug: `all-ami-updates-${process.env.AWS_ACCOUNT}`,
      branchNameStrict: "true",
      branchTopic: "{{{depNameSanitized}}}-{{{newValue}}}",
    }
  ],
}
```
works and I get "branchName": "renovate/ami_id-ami-123".

Also placing branchTopic outside the package rules works

```
module.exports = {
  ...baseConfig,
  branchTopic: "{{{depNameSanitized}}}-{{{newDigest}}}-{{{newValue}}}",

packageRules: [
    {
      matchPackageNames: ["ami-*"],
      groupName: `all AMI updates for ${process.env.AWS_ACCOUNT}`,
      groupSlug: `all-ami-updates-${process.env.AWS_ACCOUNT}`,
      branchNameStrict: "true",
    }
  ],
}
```
and i even get the digest into the branch name like = "branchName": "renovate/ami_id-amazon-eks-node-1.29-vxxxx-xxxx-ami-123"

So it feels like there is an bug/error in how the digest gets read and renovate reverts to the default wich doesnt include the patch version.

This is a "new" problem with version 38

link to mininmal reproduction:  https://github.com/Marvin-Highcreek/renovate-digest-bug-report/tree/main

### Logs (if relevant)

<details><summary>Logs</summary>

```

"deps": [
               {
                 "depName": "ami_id",
                 "packageName": "[{\"Name\":\"owner-id\",\"Values\":[\"229653807893\"]},{\"Name\":\"name\",\"Values\":[\"amazon-eks-node-1.29-*\"]}]",
                 "currentValue": "ami-123",
                 "currentDigest": "amazon-eks-node-1.xx-vxxx",
                 "datasource": "aws-machine-image",
                 "versioning": "aws-machine-image",
                 "replaceString": "  # amiFilter=[{\"Name\":\"owner-id\",\"Values\":[\"123456\"]},{\"Name\":\"name\",\"Values\":[\"amazon-eks-node-1.29-*\"]}]\n  # currentImageName=amazon-eks-node-1.xx-vxxx\n  ami_id = \"ami-123\"",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "ami-123",
                     "newValue": "ami-123",
                     "newDigest": "amazon-eks-node-1.xx-vxxx",
                     "releaseTimestamp": "XXX",
                     "newMajor": 1,
                     "newMinor": 0,
                     "newPatch": 0,
                     "updateType": "patch",
                     "branchName": "renovate/ami_id-1.x"
                   }
                 ],
                 "warnings": [],
                 "currentVersion": "ami-123",
                 "isSingleVersion": true,
                 "fixedVersion": "ami-1234"
               }
             ],
             "matchStrings": [
               ".*amiFilter=(?<packageName>.*?)\n(.*currentImageName=(?<currentDigest>.*?)\n)?(.*  \n)?.*?(?<depName>[a-zA-Z0-9-_:]*)[ ]*?[:|=][ ]*?[\"|']?(?<currentValue>ami-[a-z0-9]{17})[\"|']?.*"
             ],
             "datasourceTemplate": "aws-machine-image",
             "versioningTemplate": "aws-machine-image",
             "packageFile": "terraform/environments/foo/terragrunt.hcl"
           },

```

</details>

**Comment by viceice** [comment]
use `matchDepName` instead of `machtPackageName`, see your logs fog the different properties

**Reply by Marvin-Highcreek** [reply]
I did that here: 

![CleanShot 2024-09-17 at 17 02 45@2x](https://github.com/user-attachments/assets/033185d5-6a7a-4621-b49c-a71865748e0f)

but still got the default  branchname:

```
"updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "ami-1234",
                     "newValue": "ami-1234",
                     "newDigest": "amazon-eks-node-1.XX-123",
                     "releaseTimestamp": "2024-09-17T03:31:55.000Z",
                     "newMajor": 1,
                     "newMinor": 0,
                     "newPatch": 0,
                     "updateType": "patch",
                     "branchName": "renovate/ami_id-1.x"
                   }
                 ],
                 "warnings": [],
                 "currentVersion": "ami-123",
                 "isSingleVersion": true,
                 "fixedVersion": "ami-1234"
               }
             ],
 ```
 
 I see if I can find the time to look further into it and find the exact point of error

**Reply by viceice** [reply]
please paste the full object for this dependency
